src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,698 @@
+import asyncio
+import warnings
+from collections import OrderedDict
+from collections.abc import AsyncIterator, Iterable, Iterator
+from pathlib import Path
+from threading import Lock
+from typing import Any, cast
+
+from zarr.abc.store import (
+    ByteRequest,
+    Store,
+)
+from zarr.core.buffer import Buffer, BufferPrototype
+from zarr.core.buffer.core import default_buffer_prototype
+from zarr.storage._utils import normalize_path
+
+
+def _listdir_from_keys(store: Store, path: str) -> list[str]:
+    """
+    Extract directory listing from store keys by filtering keys that start with the given path.
+    """
+    children: set[str] = set()
+    # Handle both Store objects and dict-like objects
+    if hasattr(store, "keys") and callable(store.keys):
+        keys = [str(k) for k in store.keys()]  # Ensure keys are strings  # noqa: SIM118
+    else:
+        # For stores that don't have keys method, we can't list them
+        return []
+
+    for key in keys:
+        if key.startswith(path) and len(key) > len(path):
+            suffix = key[len(path) :]
+            child = suffix.split("/")[0]
+            children.add(child)
+    return sorted(children)
+
+
+def listdir(store: Store, path: Path | None = None) -> list[str]:
```

Comment:
This whole function can be deleted, because it's only used in `listdir()` below, which can also be delted (see other comment below).